lib/extended_json.js

```diff
@@ -265,38 +265,68 @@ function serializeValue(value, options) {
   return value;
 }
 
+const bsonTypeMappings = {
+  Binary: o => new Binary(o.value(), o.subtype),
+  Code: o => new Code(o.code, o.scope),
+  DBRef: o => new DBRef(o.collection || o.namespace, o.oid, o.db, o.fields),  // "namespace" for 1.x library backwards compat
+  Decimal128: o => new Decimal128(o.bytes),
+  Double: o => new Double(o.value),
+  Int32: o => new Int32(o.value),
+  Long: o => Long.fromBits(     // underscore variants for 1.x backwards compatibility
+    o.low != null ? o.low : o.low_,
+    o.low != null ? o.high : o.high_,
+    o.low != null ? o.unsigned : o.unsigned_
+  ),
+  MaxKey: o => new MaxKey(),
+  MinKey: o => new MinKey(),
+  ObjectID: o => new ObjectId(o),
+  ObjectId: o => new ObjectId(o), // support 4.0.0/4.0.1 before _bsontype was reverted back to ObjectID
+  BSONRegExp: o => new BSONRegExp(o.pattern, o.options),
+  Symbol: o => new Symbol(o.value),
+  Timestamp: o => Timestamp.fromBits(o.low, o.high)
+};
+
 function serializeDocument(doc, options) {
   if (doc == null || typeof doc !== 'object') throw new Error('not an object instance');
 
-  // the "document" is really just a BSON type
-  if (doc._bsontype) {
-    if (doc._bsontype === 'ObjectID') {
-      // Deprecated ObjectID class with capital "D" is still used (e.g. by 'mongodb' package). It has
-      // no "toExtendedJSON" method, so convert to new ObjectId (lowercase "d") class before serializing
-      doc = ObjectId.createFromHexString(doc.toString());
+  const bsontype = doc._bsontype;
+  if (typeof bsontype === 'undefined') {
+
+    // It's a regular object. Recursively serialize its property values. 
+    const _doc = {};
+    for (let name in doc) {
+      _doc[name] = serializeValue(doc[name], options);
     }
-    if (typeof doc.toExtendedJSON === 'function') {
-      // TODO: the two cases below mutate the original document! Bad.  I don't know
-      // enough about these two BSON types to know how to safely clone these objects, but
-      // someone who knows MongoDB better should fix this to clone instead of mutating input objects.
-      if (doc._bsontype === 'Code' && doc.scope) {
-        doc.scope = serializeDocument(doc.scope, options);
-      } else if (doc._bsontype === 'DBRef' && doc.oid) {
-        doc.oid = serializeDocument(doc.oid, options);
+    return _doc;
+
+  } else if (typeof bsontype === 'string') {
+
+    // the "document" is really just a BSON type object
+    let _doc = doc;
+    if (typeof _doc.toExtendedJSON !== 'function') {
+      // There's no EJSON serialization function on the object. It's probably an
+      // object created by a previous version of this library or (another library 
```

Comment:
nit: in the comment the `or (another library` looks like the or should be within the parens

Reply:
fixed